Report playback to PMS; closes #2

enen92 opened the pull request:
This PR implements `xbmc.Player` in mediaimporter.plex and reports/syncs the Player state to Plex Media Server.

As discussed in #2 , a thread is required to continuously update the playback time (otherwise Plex just restarts time to 0). I tried to follow the same implementation you did in mediaimporter.emby

enen92 committed:
Report playback to PMS; closes #2

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,186 @@
+class Player(xbmc.Player):
+    def __init__(self):
+        '''Initializes the player'''
+        super(xbmc.Player, self).__init__()
```
I just realized: shouldn't this be `super(Player, self).__init__()`? It would also be wrong in mediaimporter.emby...

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,186 @@
+
+    def onPlayBackEnded(self):
+        '''Triggered when playback ends. Resets the player state and inherently kills the reporting loop'''
+        self._reset()
```
Don't you need to report any specific "action"? Is not reporting any progress anymore enough?

enen92 replied:
Not in case of plex. Stop already clears everything server side

Montellese replied:
Where do you tell the Plex server that playback has stopped? IIRC the difference between `onPlayBackStopped()` and `onPlayBackEnded()` is that the former is triggered if the user manually stops playback and the latter is triggered if playback ends at the end of the video / file.

enen92 replied:
Will add

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,186 @@
+
+        if not itemId.isdigit():
+            log('invalid item id {} (non digit). Kodi will not report playback state to Plex Media Server' \
+                .format(matchingItems[0]), xbmc.LOGERROR)
```
Maybe log the specific media provider.

enen92 replied:
This was a leftover.  It is a sane check to verify the item is a digit since the plex API uses it as key (I doubt id's can be non-digits anyway) and the `updateTimeline` call expects an `int`. In the hypothetic case it is not an integer we should return early and avoid the expensive operation of getting the provider and the importer. If that operation can be made cheap, I'll add it later to the log item.

enen92 replied:
In fact I was not returning early (fixed).

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,186 @@
+            self._duration = int(self.getTotalTime() * 1000)
+
+            # start playback monitoring thread
+            if not self._playback_monitor or not self._playback_monitor.isAlive():
```
Any reason why you don't use the `Progress()` approach used in `mediaimporter.emby`? The player is already running as part of a python service which is constantly running.

enen92 replied:
sounds reasonable

enen92 replied:
To be honest I didn't get the `Process` logic at first. If the observer is a running thread it doesn't make any sense to create a new one and we can simply reuse the observer to call into the player instance. 👍 

Thinking about it. Is the implementation in mediaimporter.emby thread-safe? I am not 100% sure it is. Player events, AFAIK will be trigged by the PythonInvoker while your observer thread is also updating the same variables. It's probably better to keep a lock on the report playback method (or around any critical member variables). In python you can use contextlib to do it easily:

```python
import threading
self._lock = threading.Lock()

with self._lock:
     do_something()
```

Montellese replied:
You're right. TBH I didn't give it much thought. Adding a lock around the callback actions and the `Process()` logic makes sense.

Montellese replied:
I've created a bug at https://github.com/Montellese/mediaimporter.emby/issues/7 for `mediaimporter.emby`.

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,186 @@
+    def onAVStarted(self):
+        '''Triggered when playback actually starts'''
+        self._startPlayback()
+        self._syncPlaybackState(time=0, state=PLEX_PLAYER_PLAYING)
```
`time=0` is not correct if the video is resumed from a specific position.

enen92 replied:
good catch

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,186 @@
+
+    def _getPlayingTime(self):
+        '''Gets current xbmc.Player time in miliseconds'''
+        return int(self.getTime() * 1000)
```
This conversion to milliseconds appears twice in this code. Maybe add a utility method to `plex.api.Api` (similar to `mediaimporter.emby`).

enen92 replied:
decided to add to lib.utils since it sits in between kodi and plex

Montellese reviewed (changes requested):
Looks good but I'm wondering whether we should use the separate thread approach or the `Process()` approach used in `mediaimporter.emby`. I have no experience with threading in python...

enen92 commented on plex/player.py:
```diff
@@ -0,0 +1,186 @@
+
+        self._itemId = int(itemId)
+
+        for mediaProvider in self._providers.values():
```
I don't really like this approach as it makes you iterate over all the imported items from every provider. Isn't there an alternative implementation  to get it directly from the mediaimport API?

Montellese replied:
How would such a call look like? You only have a file or a list item. The fact that the identification happens over the unique ID is specific to the media importer implementation. Another media importer could use a different approach (e.g. determine the unique identifier from the file name). Furthermore you have no idea from which media provider / import the item is coming from.

I'd have to check if we could make the information from which media provider an item has been imported available on the list item. Then you would only have to retrieve all imported items from the media provider matching the item's assign media provider.

enen92 replied:
Hum can you set both the media provider id and the media importer id as listitem properties? That way we could from python just check if both properties are defined and later query the mediaimport API for the specific provider/importer objects providing the Id

Montellese replied:
I could provide APIs to get the media provider identifier and the media import path through the `ListItem`'s class. Then you can simply use `self._providers[item.getMediaProviderId()]`.

Montellese replied:
I just realized that the `xbmc.Player` API does not give us access to the `ListItem` so that would have to be added as well. I've created an improvement at https://github.com/Montellese/xbmc/issues/60 but IMO we should not block this PR due to this. Maybe add a TODO comment that this should be improved.

enen92 replied:
That would be awesome since there's a pending issue in Kodi that to be solved needs the Listitem to be accessed from the player (json-rpc). I tried my luck with that issue but it seems that it has to be implemented in ApplicationPlayer and then proxied to every single player (as far I as could tell). I failed miserably doing it for retroplayer and UPnPPlayer (they don't hold a listitem). PAPlayer and VideoPlayer seem to be easy.

I'll get back to this PR as soon as possible

Montellese replied:
I've fixed this in 70ee52452b9382d6f63fb492651cf4e3f3354b22 after I added `Player.getPlayingItem()` and `ListItem.getMediaProviderId()` to Kodi's python API. That way we can directly lookup the matching media provider from `self._providers` using the item's media provider ID.

Montellese commented on lib/utils.py:
```diff
@@ -52,6 +52,10 @@ def localise(id):
     return normalizeString(__addon__.getLocalizedString(id))
 
 
+def toMiliseconds(seconds):
```
```suggestion
def toMilliseconds(seconds):
```

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,204 @@
+from plex.constants import PLEX_PROTOCOL, PLEX_PLAYER_PLAYING, PLEX_PLAYER_PAUSED, PLEX_PLAYER_STOPPED
+from plex.server import Server
+
+from lib.utils import log, mediaProvider2str, toMiliseconds
```
```suggestion
from lib.utils import log, mediaProvider2str, toMilliseconds
```

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,204 @@
+                self._itemId,
+                Api.getPlexMediaClassFromMediaType(videoInfoTag.getMediaType())
+            )
+            self._duration = toMiliseconds(self.getTotalTime())
```
```suggestion
            self._duration = toMilliseconds(self.getTotalTime())
```

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,204 @@
+
+    def _getPlayingTime(self):
+        '''Gets current xbmc.Player time in miliseconds'''
+        return toMiliseconds(self.getTime())
```
```suggestion
        return toMilliseconds(self.getTime())
```

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,204 @@
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        self._providers[mediaProvider.getIdentifier()] = mediaProvider
```
This needs locking as well.

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,204 @@
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        del self._providers[mediaProvider.getIdentifier()]
```
This needs locking as well.

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,204 @@
+
+    def Process(self):
+        '''Called from the observer thread to periodically report the state to PMS'''
+        lastreport = self._state.get('lastreport')
```
This needs locking as well.

Montellese reviewed (changes requested):
Looks good apart from the typo and from the missing locks. I haven't commented on all missing locks though.

enen92 commented:
Hey @Montellese did the changes you suggested and rebased on master. Should be good now.
I faced a few issues with locking caused by acquiring  the lock on the same thread multiple times (e.g. `_reset` is acquiring the lock and can be called from multiple threads but can also be called in `_startPlayback` which also holds the lock. For this reason I decided to move to the `RLock` (reentrant lock interface). Its pretty more robust than the standard `Lock` since the same thread can obtain the lock multiple times while another thread must wait for all the computation performed by the holding thread to finish.

Montellese commented on plex/player.py:
```diff
@@ -0,0 +1,218 @@
+
+    def onPlayBackStarted(self):
+        '''Triggered when xbmc.Player is started'''
+        self._reset()
```
I would take the lock here to make sure that no other thread can come between executing `self._reset()` and `self._getPlayingFile()`. Otherwise you might end up with this thread executing `self._reset()` then another thread does something and then this thread executes `self._getPlayingFile()` which might again end up with an invalid state.

The same applies to all other methods. Then you can get rid of the locks in the private methods and use a normal `Lock` again.

Montellese reviewed (changes requested):
Thanks for making the changes. I actually considered using an `RLock` as well in `mediaimporter.emby` but ended up using a normal `Lock` in every "public" method, see my other review comment.

enen92 commented:
Ok so you prefer the locks on the public methods instead of private ones. I implemented them in private methods because I somehow though you'd prefer it that way (hides implementation details). 

Moved them to public methods now and used the simple `Lock` again. Tested it for a while and it's working. Now it should be good to go I guess.
Thanks for the review

Montellese reviewed (approved):
Thanks for your support.

There's no use in hiding the implementation detail if it is not implemented correctly ;-)
